src/pm/util/pmiport.c

```diff
@@ -367,54 +384,63 @@ int MPIE_ConnectToPort(char *hostname, int portnum)
     }
-    hp = gethostbyname(hostname);
-    if (!hp) {
+
+    memset(&hints, 0, sizeof(struct addrinfo));
+    hints.ai_family = AF_UNSPEC;
+    hints.ai_socktype = SOCK_STREAM;
+    hints.ai_flags = 0;
+    hints.ai_protocol = 0;
+    snprintf(port, 16, "%d", portnum);
+    status = getaddrinfo(hostname, port, &hints, &result);
+    if (status != 0) {
         return -1;
     }
-
-    memset((void *) &sa, 0, sizeof(sa));
+    //memset((void *) &sa, 0, sizeof(sa));
     /* POSIX might define h_addr_list only and not define h_addr */
-#ifdef HAVE_H_ADDR_LIST
+/*#ifdef HAVE_H_ADDR_LIST
     memcpy((void *) &sa.sin_addr, (void *) hp->h_addr_list[0], hp->h_length);
 #else
     memcpy((void *) &sa.sin_addr, (void *) hp->h_addr, hp->h_length);
 #endif
     sa.sin_family = hp->h_addrtype;
     sa.sin_port = htons((unsigned short) portnum);
+	*/
+    for (rp = result; rp != NULL; rp = rp->ai_next) {
+        connect_failed = 0;
+        fd = socket(rp->ai_family, rp->ai_socktype, rp->ai_protocol);
+        if (fd < 0) {
+            continue;
+        }
 
-    fd = socket(AF_INET, SOCK_STREAM, TCP);
-    if (fd < 0) {
-        return -1;
-    }
-
-    if (setsockopt(fd, IPPROTO_TCP, TCP_NODELAY, (char *) &optval, sizeof(optval))) {
-        perror("Error calling setsockopt:");
-    }
-
-    /* We wait here for the connection to succeed */
-    if (connect(fd, (struct sockaddr *) &sa, sizeof(sa)) < 0) {
-        switch (errno) {
-            case ECONNREFUSED:
-                /* (close socket, get new socket, try again) */
-                if (q_wait)
-                    close(fd);
-                return -1;
-
-            case EINPROGRESS:  /*  (nonblocking) - select for writing. */
-                break;
-
-            case EISCONN:      /*  (already connected) */
-                break;
-
-            case ETIMEDOUT:    /* timed out */
-                return -1;
+        if (setsockopt(fd, IPPROTO_TCP, TCP_NODELAY, (char *) &optval, sizeof(optval))) {
+            perror("Error calling setsockopt:");
+        }
 
-            default:
-                return -1;
+        /* We wait here for the connection to succeed */
+        if (connect(fd, rp->ai_addr, rp->ai_addrlen) != -1) {   //Success
+            break;
+        } else {        //Fail
+            switch (errno) {
+                case ECONNREFUSED:
+                    /* (close socket, get new socket, try again) */
+                    if (q_wait)
+                        close(fd);
+                    connect_failed = 1;
+                    break;
+                case EINPROGRESS:      /*  (nonblocking) - select for writing. */
+                    break;
+                case EISCONN:  /*  (already connected) */
+                    break;
+                case ETIMEDOUT:        /* timed out */
+                    connect_failed = 1;
+                    break;
+                default:
+                    connect_failed = 1;
+                    break;
+            }
```

Comment:
You mean that we should prioritize IPV4 first, if all IPV4 failed, then we try with IPV6, right?

Reply:
Hi, Yanfei, I think as you said, if the server is listening on the IPV4 socket and rp is IPV6, connect is failed, then we mark the connect_failed = 1 and we break the switch (not break the for loop) and proceed to the next rp, once any one of the rp is get connected successfully we break from the loop and return. On the other side, if all of the connect failed, connect_failed must be 1, then we return -1 instead.

Reply:
I don't think so. The loopis iterating over each `rp`, but the loop is broken once any of this rp is connected or failed to connect. The problem is that when a host has two (or more) IP addresses (regardless v4 or v6), the loop would break if any of these addresses failed to connect without trying the other addresses left, if there is any).

Reply:
when a rp is connected, yes it will break the loop and return the fd. But if the rp is failed to connect, the code will enter switch-case and each break under each case will just break from the switch case, not break from the loop. The loop will continues